### PR TITLE
Implemented all requirements.

### PR DESCRIPTION
**Pull request by cudabanger**

### CODE REVIEW

File: Car.h [hunk 1]
```diff
@@ -1,9 +1,13 @@
 #pragma once
+#include "Vehicle.h"
 
 struct Car : public Vehicle
```
Comment (matkatmusic) [inline Car.h]:
```    
    10) correct any unnecessary keywords related to access specifiers used in the UDTs.
```
- [x] revise

File: Highway.cpp [hunk 2]
```diff
@@ -13,22 +16,53 @@ void Highway::changeSpeed(int newSpeed)
...
+    if ( Motorcycle* bike = dynamic_cast<Motorcycle*>(v))
+    {
+        // the member function that doesn't evade the cops
+        bike->lanesplitAndRace(speedLimit);
+    }
+
+    if ( SemiTruck* truck = dynamic_cast<SemiTruck*>(v))
+    {
+        truck->setSpeed(speedLimit);
+    }
```
Comment (matkatmusic) [inline Highway.cpp]: 1) use `else if`.  you're currently checking if `v` is a `Motorcycle*`, even if it's already been determined to be a `Car*`.
2) you need to write a function in `SemiTruck`.  Don't call `setSpeed` here.

- [x] revise

File: Highway.cpp [hunk 3]
```diff
@@ -13,22 +16,53 @@ void Highway::changeSpeed(int newSpeed)
...
+
+    if (Motorcycle* bike = dynamic_cast<Motorcycle*>(v))
+    {
+        bike->tryToEvade();
+    }
+
+    if (SemiTruck* truck = dynamic_cast<SemiTruck*>(v))
+    {
+        truck->tryToEvade();
+    }
```
Comment (matkatmusic) [inline Highway.cpp]: 1) use `else if`. 
2) "trucks pull over".   Think back to Project 1, where you came up with plain-english expressions and wrote pseudo-code that did the same thing.   Apply that concept here.
- [x] revise

File: HighwayPatrol.cpp [hunk 4]
```diff
@@ -1,11 +1,18 @@
...
 #include <iostream>
+#include "Highway.h"
 
 HighwayPatrol::HighwayPatrol() : Vehicle("HighwayPatrol")
 {
 
 }
 
+// I need one out-of-line virtual function definition to avoid weak vtables warning
+HighwayPatrol::~HighwayPatrol()
```
Comment (matkatmusic) [inline HighwayPatrol.cpp:15]: Correct!

File: HighwayPatrol.h [hunk 5]
```diff
@@ -2,10 +2,15 @@
 
 #include "Vehicle.h"
 
-struct HighwayPatrol : Vehicle 
+struct Highway;
+struct HighwayPatrol : public Vehicle
```
Comment (matkatmusic) [inline HighwayPatrol.h]:
```    
    10) correct any unnecessary keywords related to access specifiers used in the UDTs.
```
- [x] revise

File: Motorcycle.h [hunk 6]
```diff
@@ -2,10 +2,13 @@
 
 #include <iostream>
 #include <limits>
+#include "Vehicle.h"
 
 struct Motorcycle : public Vehicle
```
Comment (matkatmusic) [inline Motorcycle.h]:
```    
    10) correct any unnecessary keywords related to access specifiers used in the UDTs.
```
- [x] revise

File: SemiTruck.cpp [hunk 7]
```diff
@@ -0,0 +1,14 @@
+#include "SemiTruck.h"
+
+SemiTruck::SemiTruck(const std::string& n) : Vehicle(n) {}
+
+void SemiTruck::tryToEvade()
+{
+    setSpeed(0);
+    std::cout << name << ": hello officer, what seems to be the problem?" << std::endl;
+}
```
Comment (matkatmusic) [inline SemiTruck.cpp]: The internals of this function are the exact opposite of `trying to evade`. 
name things what they do.  This truck has pulled over.
- [x] revise

File: SemiTruck.cpp [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+void SemiTruck::setSpeed(int speed)
+{
+    Vehicle::setSpeed(speed);    
+}
```
Comment (matkatmusic) [inline SemiTruck.cpp]: it's pointless to override the base class function and only call the base class function from the derived function.  The only thing this does is add more function calls to the call stack, which adds bloat in the compiled binary. 
```
   5) Add a Truck type
        semi-trucks don't evade when they're pulled over, unlike cars and motorcycles.  
        They'll also need a member function that can be called when they are added to the highway.  Look at the Highway member functions for more hints on this.
```
- [x] revise

File: SemiTruck.h [hunk 9]
```diff
@@ -1,2 +1,12 @@
 #pragma once
+#include "Vehicle.h"
 
+struct SemiTruck : public Vehicle
```
Comment (matkatmusic) [inline SemiTruck.h]:
```
  10) correct any unnecessary keywords related to access specifiers used in the UDTs
```
- [x] revise

File: SemiTruck.h [hunk 10]
```diff
@@ -1,2 +1,12 @@
...
+{
+    SemiTruck(const std::string& s);
+    SemiTruck(const SemiTruck&) = default;
+    SemiTruck& operator=(const SemiTruck&) = default;
+
+    void tryToEvade() override;
+    void setSpeed(int s) override;
```
Comment (matkatmusic) [inline SemiTruck.h]:
```
        semi-trucks don't evade when they're pulled over, unlike cars and motorcycles.  
        They'll also need a member function that can be called when they are added to the highway.  Look at the Highway member functions for more hints on this.
```
- [x] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]
In addition to the things I've marked below, You didn't implement `HighwayPatrol::30`

File: HighwayPatrol.cpp (hunk 4 above)
Comment (cudabanger) [reply HighwayPatrol.cpp:15]: I don't get it.  Are you acknowledging my comment?  The code following it *is* the out-of-line code so what is it that you want me to do here?

File: HighwayPatrol.cpp [hunk 11]
```diff
@@ -28,7 +35,7 @@ void HighwayPatrol::pullOver( Vehicle* v, bool willArrest, Highway* h )
     if( willArrest )
     {
         //print the vehicle type in this std::cout between "THE [" and "] PULL". 
-        std::cout << name << ": YOU IN THE [ " << " ] PULL OVER AND SHOW YOUR HANDS" << std::endl;
+        std::cout << name << ": YOU IN THE " << v->name << " PULL OVER AND SHOW YOUR HANDS" << std::endl;
```
Comment (matkatmusic) [inline HighwayPatrol.cpp]: `v->name` is not the `vehicle type`.
- [x] revise

File: SemiTruck.cpp [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+void SemiTruck::pullOver()
+{
+    setSpeed(0);
+    std::cout << name << ": hello officer, what seems to be the problem?" << std::endl;
+}
+
+void SemiTruck::getOnHighway(int speed)
+{
+    Vehicle::setSpeed(speed);    
+}
```
Comment (matkatmusic) [inline SemiTruck.cpp]: add a `std::cout` statement to this function. 
- [x] revise

**Review by matkatmusic — CHANGES_REQUESTED** [review]

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**